src/NHibernate/Loader/Loader.cs

```diff
@@ -1708,60 +1716,54 @@ private CacheableResultTransformer CreateCacheableResultTransformer(QueryParamet
 				queryParameters.HasAutoDiscoverScalarTypes, SqlString);
 		}
 
-		internal IList GetResultFromQueryCache(ISessionImplementor session, QueryParameters queryParameters,
-											  ISet<string> querySpaces, IQueryCache queryCache,
-											  QueryKey key)
+		internal bool CanGetFromCache(ISessionImplementor session, QueryParameters queryParameters)
 		{
-			IList result = null;
-
-			if (!queryParameters.ForceCacheRefresh && session.CacheMode.HasFlag(CacheMode.Get))
-			{
-				IPersistenceContext persistenceContext = session.PersistenceContext;
+			return !queryParameters.ForceCacheRefresh && session.CacheMode.HasFlag(CacheMode.Get);
+		}
 
-				bool defaultReadOnlyOrig = persistenceContext.DefaultReadOnly;
+		internal IList GetResultFromQueryCache(
+			ISessionImplementor session, QueryParameters queryParameters, ISet<string> querySpaces,
+			IQueryCache queryCache, QueryKey key)
+		{
+			if (!CanGetFromCache(session, queryParameters))
+				return null;
 
-				if (queryParameters.IsReadOnlyInitialized)
-					persistenceContext.DefaultReadOnly = queryParameters.ReadOnly;
-				else
-					queryParameters.ReadOnly = persistenceContext.DefaultReadOnly;
+			var result = queryCache.Get(
+				key, queryParameters, 
+				queryParameters.HasAutoDiscoverScalarTypes
+					? null
+					: key.ResultTransformer.GetCachedResultTypes(ResultTypes),
+				querySpaces, session);
 
-				try
+			if (_factory.Statistics.IsStatisticsEnabled)
+			{
```

Comment:
I was tempted to move the statistics handling on the query cache side too, but that is not done for concurrency strategy caches either. So for consistency, I have kept it here.